File: clients/java/src/main/java/io/merklekv/client/Benchmark.java
package io.merklekv.client;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;

import java.util.*;
import java.util.concurrent.ThreadLocalRandom;

public class Benchmark {
    private static final String DEFAULT_HOST = "localhost";
    private static final int DEFAULT_PORT = 7379;
    private static final int DEFAULT_WARMUP_OPS = 5000;
    private static final int DEFAULT_WORKLOAD_OPS = 100000;
    private static final int[] DEFAULT_PIPELINE_DEPTHS = {1, 8, 32, 128};
    private static final int DEFAULT_KEYSPACE_SIZE = 10000;
    private static final int DEFAULT_VALUE_SIZE = 16;
    private static final double DEFAULT_GET_SET_RATIO = 0.5;

    public static void main(String[] args) throws Exception {
        BenchmarkConfig config = parseArgs(args);
        
        System.out.println("Java MerkleKV Client Benchmark");
        System.out.println("==============================");
        System.out.println("Host: " + config.host + ":" + config.port);
        System.out.println("Warmup: " + config.warmupOps + " ops");
        System.out.println("Workload: " + config.workloadOps + " ops");
        System.out.println("Pipeline depths: " + Arrays.toString(config.pipelineDepths));
        System.out.println("Keyspace: " + config.keyspaceSize + " keys");
        System.out.println("Value size: " + config.valueSize + " bytes");
        System.out.println();

        List<BenchmarkResult> results = runBenchmark(config);
        
        // Print results
        System.out.println("Results:");
        System.out.println("========");
        for (BenchmarkResult result : results) {
            System.out.printf("Operation: %s (depth=%d)%n", result.operation, result.depth);
            System.out.printf("  P50: %.2f ms%n", result.p50Latency);
            System.out.printf("  P95: %.2f ms%n", result.p95Latency);
            System.out.printf("  P99: %.2f ms%n", result.p99Latency);
            System.out.printf("  Throughput: %.1f ops/sec%n", result.throughput);
            System.out.printf("  Errors: %d/%d%n", result.errors, result.totalOps);
            System.out.println();
        }

        // Output JSON for CI parsing
        ObjectMapper mapper = new ObjectMapper();
        ArrayNode jsonResults = mapper.createArrayNode();
        for (BenchmarkResult result : results) {
            ObjectNode resultNode = mapper.createObjectNode();
            resultNode.put("operation", result.operation);
            resultNode.put("depth", result.depth);
            resultNode.put("p50_ms", result.p50Latency);
            resultNode.put("p95_ms", result.p95Latency);
            resultNode.put("p99_ms", result.p99Latency);
            resultNode.put("throughput_ops_sec", result.throughput);
            resultNode.put("total_ops", result.totalOps);
            resultNode.put("errors", result.errors);
            jsonResults.add(resultNode);
        }
        System.out.println("JSON Results:");
        System.out.println(mapper.writeValueAsString(jsonResults));

        // Check p50 compliance (< 5ms)
        double maxP50 = results.stream().mapToDouble(r -> r.p50Latency).max().orElse(0.0);
        if (maxP50 < 5.0) {
            System.out.println("Performance Gate: ✅ PASS (max p50: " + String.format("%.2f", maxP50) + " ms)");
            System.exit(0);
        } else {
            System.out.println("Performance Gate: ❌ FAIL (max p50: " + String.format("%.2f", maxP50) + " ms)");
            System.exit(1);
        }
    }

    private static BenchmarkConfig parseArgs(String[] args) {
        BenchmarkConfig config = new BenchmarkConfig();
        config.host = DEFAULT_HOST;
        config.port = DEFAULT_PORT;
        config.warmupOps = DEFAULT_WARMUP_OPS;
        config.workloadOps = DEFAULT_WORKLOAD_OPS;
        config.pipelineDepths = DEFAULT_PIPELINE_DEPTHS.clone();
        config.keyspaceSize = DEFAULT_KEYSPACE_SIZE;
        config.valueSize = DEFAULT_VALUE_SIZE;
        config.getSetRatio = DEFAULT_GET_SET_RATIO;

        for (int i = 0; i < args.length - 1; i++) {
            switch (args[i]) {
                case "-host":
                    config.host = args[i + 1];
                    break;
                case "-port":
                    config.port = Integer.parseInt(args[i + 1]);
                    break;
                case "-ops":
                    config.workloadOps = Integer.parseInt(args[i + 1]);
                    break;
                case "-warmup":
                    config.warmupOps = Integer.parseInt(args[i + 1]);
                    break;
                case "-depths":
                    String[] depthStrs = args[i + 1].split(",");
                    config.pipelineDepths = new int[depthStrs.length];
                    for (int j = 0; j < depthStrs.length; j++) {
                        config.pipelineDepths[j] = Integer.parseInt(depthStrs[j].trim());
                    }
                    break;
            }
        }
        return config;
    }

    private static List<BenchmarkResult> runBenchmark(BenchmarkConfig config) throws Exception {
        List<BenchmarkResult> results = new ArrayList<>();

        try (MerkleKVClient client = new MerkleKVClient(config.host, config.port)) {
            System.out.println("Connected to MerkleKV server");

            // Warmup
            System.out.println("Running warmup (" + config.warmupOps + " ops)...");
            warmup(client, config);

            // Test each pipeline depth
            for (int depth : config.pipelineDepths) {
                System.out.println("Benchmarking pipeline depth " + depth + "...");
                BenchmarkResult result = benchmarkPipelineDepth(client, config, depth);
                results.add(result);
            }
        }

        return results;
    }

    private static void warmup(MerkleKVClient client, BenchmarkConfig config) throws Exception {
        String testValue = "x".repeat(config.valueSize);
        List<String> commands = new ArrayList<>();
        
        for (int i = 0; i < config.warmupOps; i++) {
            String key = "warmup_key_" + ThreadLocalRandom.current().nextInt(config.keyspaceSize);
            
            if (ThreadLocalRandom.current().nextDouble() < config.getSetRatio) {
                commands.add("GET " + key);
            } else {
                commands.add("SET " + key + " " + testValue);
            }
            
            // Execute in batches of 100
            if (commands.size() >= 100 || i == config.warmupOps - 1) {
                client.pipeline(commands);
                commands.clear();
            }
        }
    }

    private static BenchmarkResult benchmarkPipelineDepth(MerkleKVClient client, BenchmarkConfig config, int depth) throws Exception {
        List<Double> latencies = new ArrayList<>();
        int totalOps = 0;
        int errors = 0;
        String testValue = "x".repeat(config.valueSize);

        long start = System.nanoTime();

        while (totalOps < config.workloadOps) {
            // Prepare batch of commands
            int batchSize = Math.min(depth, config.workloadOps - totalOps);
            List<String> commands = new ArrayList<>(batchSize);
            
            for (int i = 0; i < batchSize; i++) {
                String key = "bench_key_" + ThreadLocalRandom.current().nextInt(config.keyspaceSize);
                
                if (ThreadLocalRandom.current().nextDouble() < config.getSetRatio) {
                    commands.add("GET " + key);
                } else {
                    commands.add("SET " + key + " " + testValue);
                }
            }

            // Measure pipeline operation
            long opStart = System.nanoTime();
            try {
                client.pipeline(commands);
                long opDuration = System.nanoTime() - opStart;
                
                // Use total latency and divide by depth for per-operation latency
                double totalLatencyMs = opDuration / 1_000_000.0;
                latencies.add(totalLatencyMs);
            } catch (Exception e) {
                errors++;
            }

            totalOps += batchSize;
        }

        long duration = System.nanoTime() - start;
        double throughput = (double) totalOps / (duration / 1_000_000_000.0);

        // Calculate percentiles - convert to per-operation latency
        Collections.sort(latencies);
        List<Double> perOpLatencies = new ArrayList<>();
        for (double totalLatency : latencies) {
            perOpLatencies.add(totalLatency / depth);
        }
        Collections.sort(perOpLatencies);

        double p50 = perOpLatencies.isEmpty() ? 0.0 : perOpLatencies.get(perOpLatencies.size() * 50 / 100);
        double p95 = perOpLatencies.isEmpty() ? 0.0 : perOpLatencies.get(perOpLatencies.size() * 95 / 100);
        double p99 = perOpLatencies.isEmpty() ? 0.0 : perOpLatencies.get(perOpLatencies.size() * 99 / 100);

        return new BenchmarkResult("Pipeline", depth, p50, p95, p99, throughput, totalOps, errors);
    }

    static class BenchmarkConfig {
        String host;
        int port;
        int warmupOps;
        int workloadOps;
        int[] pipelineDepths;
        int keyspaceSize;
        int valueSize;
        double getSetRatio;
    }

    static class BenchmarkResult {
        final String operation;
        final int depth;
        final double p50Latency;
        final double p95Latency;
        final double p99Latency;
        final double throughput;
        final int totalOps;
        final int errors;

        BenchmarkResult(String operation, int depth, double p50Latency, double p95Latency, double p99Latency,
                       double throughput, int totalOps, int errors) {
            this.operation = operation;
            this.depth = depth;
            this.p50Latency = p50Latency;
            this.p95Latency = p95Latency;
            this.p99Latency = p99Latency;
            this.throughput = throughput;
            this.totalOps = totalOps;
            this.errors = errors;
        }
    }
}

File: clients/rust/src/bin/benchmark.rs
use std::time::Instant;
use std::env;
use merklekv_client::{Client, Result};
use serde_json::json;

#[derive(Clone)]
struct BenchmarkConfig {
    host: String,
    port: u16,
    warmup_ops: usize,
    workload_ops: usize,
    pipeline_depths: Vec<usize>,
    keyspace_size: usize,
    value_size: usize,
    get_set_ratio: f64,
}

struct BenchmarkResult {
    operation: String,
    depth: usize,
    p50_latency: f64,
    p95_latency: f64,
    p99_latency: f64,
    throughput: f64,
    errors: usize,
    total_ops: usize,
}

fn main() -> Result<()> {
    env_logger::init();

    let mut config = BenchmarkConfig {
        host: "localhost".to_string(),
        port: 7379,
        warmup_ops: 5000,
        workload_ops: 100000,
        pipeline_depths: vec![1, 8, 32, 128],
        keyspace_size: 10000,
        value_size: 16,
        get_set_ratio: 0.5,
    };

    // Parse command line arguments
    let args: Vec<String> = env::args().collect();
    for i in 0..args.len() {
        match args.get(i).map(|s| s.as_str()) {
            Some("-host") => {
                if let Some(host) = args.get(i + 1) {
                    config.host = host.clone();
                }
            }
            Some("-port") => {
                if let Some(port) = args.get(i + 1) {
                    config.port = port.parse().unwrap_or(7379);
                }
            }
            Some("-ops") => {
                if let Some(ops) = args.get(i + 1) {
                    config.workload_ops = ops.parse().unwrap_or(100000);
                }
            }
            Some("-warmup") => {
                if let Some(warmup) = args.get(i + 1) {
                    config.warmup_ops = warmup.parse().unwrap_or(5000);
                }
            }
            Some("-depths") => {
                if let Some(depths_str) = args.get(i + 1) {
                    config.pipeline_depths = depths_str
                        .split(',')
                        .filter_map(|s| s.trim().parse().ok())
                        .collect();
                }
            }
            _ => {}
        }
    }

    println!("Rust MerkleKV Client Benchmark");
    println!("==============================");
    println!("Host: {}:{}", config.host, config.port);
    println!("Warmup: {} ops", config.warmup_ops);
    println!("Workload: {} ops", config.workload_ops);
    println!("Pipeline depths: {:?}", config.pipeline_depths);
    println!("Keyspace: {} keys", config.keyspace_size);
    println!("Value size: {} bytes", config.value_size);
    println!();

    // Run benchmark
    let results = run_benchmark(config)?;

    // Print results
    println!("Results:");
    println!("========");
    for result in &results {
        println!("Operation: {} (depth={})", result.operation, result.depth);
        println!("  P50: {:.2} ms", result.p50_latency);
        println!("  P95: {:.2} ms", result.p95_latency);
        println!("  P99: {:.2} ms", result.p99_latency);
        println!("  Throughput: {:.1} ops/sec", result.throughput);
        println!("  Errors: {}/{}", result.errors, result.total_ops);
        println!();
    }

    // Output JSON for CI parsing
    let json_results: serde_json::Value = json!(
        results.iter().map(|r| json!({
            "operation": r.operation,
            "depth": r.depth,
            "p50_ms": r.p50_latency,
            "p95_ms": r.p95_latency,
            "p99_ms": r.p99_latency,
            "throughput_ops_sec": r.throughput,
            "errors": r.errors,
            "total_ops": r.total_ops
        })).collect::<Vec<_>>()
    );
    
    println!("JSON Results:");
    println!("{}", serde_json::to_string_pretty(&json_results).unwrap());

    // Check p50 compliance (< 5ms)
    let max_p50 = results.iter()
        .map(|r| r.p50_latency)
        .fold(0.0, f64::max);

    print!("Performance Gate: ");
    if max_p50 < 5.0 {
        println!("✅ PASS (max p50: {:.2} ms)", max_p50);
        std::process::exit(0);
    } else {
        println!("❌ FAIL (max p50: {:.2} ms)", max_p50);
        std::process::exit(1);
    }
}

fn run_benchmark(config: BenchmarkConfig) -> Result<Vec<BenchmarkResult>> {
    let addr = format!("{}:{}", config.host, config.port);
    let mut client = Client::connect(&addr)?;

    // Test connectivity
    client.health_check()?;
    println!("Connected to MerkleKV server");

    // Generate test value
    let test_value = "x".repeat(config.value_size);

    let mut results = Vec::new();

    // Run warmup
    println!("Running warmup ({} ops)...", config.warmup_ops);
    run_warmup(&mut client, &config, &test_value)?;

    // Benchmark each pipeline depth
    for &depth in &config.pipeline_depths {
        println!("Benchmarking pipeline depth {}...", depth);
        
        let result = benchmark_pipeline_depth(&mut client, &config, &test_value, depth)?;
        results.push(result);
    }

    Ok(results)
}

fn run_warmup(client: &mut Client, config: &BenchmarkConfig, test_value: &str) -> Result<()> {
    let mut commands = Vec::new();
    
    for i in 0..config.warmup_ops {
        let key = format!("warmup_key_{}", i % config.keyspace_size);
        
        if (i as f64 / config.warmup_ops as f64) < config.get_set_ratio {
            commands.push(format!("GET {}", key));
        } else {
            commands.push(format!("SET {} {}", key, test_value));
        }
        
        // Execute in batches of 100
        if commands.len() >= 100 || i == config.warmup_ops - 1 {
            let _ = client.pipeline(commands.clone());
            commands.clear();
        }
    }
    
    Ok(())
}

fn benchmark_pipeline_depth(
    client: &mut Client, 
    config: &BenchmarkConfig, 
    test_value: &str, 
    depth: usize
) -> Result<BenchmarkResult> {
    let mut latencies = Vec::new();
    let mut total_ops = 0;
    let mut errors = 0;

    let start = Instant::now();

    while total_ops < config.workload_ops {
        // Prepare batch of commands
        let batch_size = depth.min(config.workload_ops - total_ops);
        let mut commands = Vec::with_capacity(batch_size);
        
        for i in 0..batch_size {
            let key = format!("bench_key_{}", (total_ops + i) % config.keyspace_size);
            
            if (i as f64 / batch_size as f64) < config.get_set_ratio {
                commands.push(format!("GET {}", key));
            } else {
                commands.push(format!("SET {} {}", key, test_value));
            }
        }

        // Measure pipeline operation
        let op_start = Instant::now();
        match client.pipeline(commands) {
            Ok(_) => {
                let op_duration = op_start.elapsed();
                // Use actual total latency for pipeline operations, similar to Go client
                latencies.push(op_duration.as_secs_f64() * 1000.0);
            }
            Err(_) => {
                errors += 1;
            }
        }

        total_ops += batch_size;
    }

    let duration = start.elapsed();
    let throughput = total_ops as f64 / duration.as_secs_f64();

    // Calculate percentiles - convert to per-operation latency for pipeline
    latencies.sort_by(|a, b| a.partial_cmp(b).unwrap());
    
    let per_op_latencies: Vec<f64> = latencies.iter()
        .map(|&total_latency| total_latency / depth as f64)
        .collect();
    
    let p50 = if per_op_latencies.is_empty() { 0.0 } else { per_op_latencies[per_op_latencies.len() * 50 / 100] };
    let p95 = if per_op_latencies.is_empty() { 0.0 } else { per_op_latencies[per_op_latencies.len() * 95 / 100] };
    let p99 = if per_op_latencies.is_empty() { 0.0 } else { per_op_latencies[per_op_latencies.len() * 99 / 100] };

    Ok(BenchmarkResult {
        operation: "Pipeline".to_string(),
        depth,
        p50_latency: p50,
        p95_latency: p95,
        p99_latency: p99,
        throughput,
        errors,
        total_ops,
    })
}
